graphics/src/AssimpLoader.cc

```diff
@@ -0,0 +1,733 @@
+  /// \param[in] _parent the parent skeleton node
+  /// \param[in] _transform the transform of the current node
+  /// \param[in] _boneNames set of bone names, used to skip nodes without a bone
+  public: void RecursiveSkeletonCreate(
+          const aiNode* _node,
+          SkeletonNode* _parent,
+          const math::Matrix4d& _transform,
+          const std::unordered_set<std::string> &_boneNames) const;
+
+  /// \brief Recursively store the bone names starting from the root node
+  /// to make sure that only nodes that map to a bone are added to the skeleton
+  /// \param[in] _scene the assimp scene
+  /// \param[in] _node the node being processed
+  /// \param[out] _boneNames set of bone names populated while recursing
+  public: void RecursiveStoreBoneNames(
+          const aiScene *_scene,
+          const aiNode* _node,
+          std::unordered_set<std::string>& _boneNames) const;
+
+  /// \brief Apply the the inv bind transform to the skeleton pose.
+  /// \remarks have to set the model transforms starting from the root in
+  /// breadth first order. Because setting the model transform also updates
+  /// the transform based on the parent's inv model transform. Setting the
+  /// child before the parent results in the child's transform being
+  /// calculated from the "old" parent model transform.
+  /// \param[in] _skeleton the skeleton to work on
+  public: void ApplyInvBindTransform(SkeletonPtr _skeleton) const;
+};
+
+//////////////////////////////////////////////////
+// Utility function to convert to std::string from aiString
+static std::string ToString(const aiString& str)
+{
+  return std::string(str.C_Str());
+}
+
+//////////////////////////////////////////////////
+math::Color AssimpLoader::Implementation::ConvertColor(aiColor4D& _color) const
+{
+  math::Color col(_color.r, _color.g, _color.b, _color.a);
+  return col;
+}
+
+//////////////////////////////////////////////////
+math::Matrix4d AssimpLoader::Implementation::ConvertTransform(
+    const aiMatrix4x4& _sm) const
+{
+  return math::Matrix4d(
+      _sm.a1, _sm.a2, _sm.a3, _sm.a4,
+      _sm.b1, _sm.b2, _sm.b3, _sm.b4,
+      _sm.c1, _sm.c2, _sm.c3, _sm.c4,
+      _sm.d1, _sm.d2, _sm.d3, _sm.d4);
+}
+
+//////////////////////////////////////////////////
+void AssimpLoader::Implementation::RecursiveCreate(const aiScene* _scene,
+    const aiNode* _node, const math::Matrix4d& _transform, Mesh* _mesh) const
+{
+  if (!_node)
+    return;
+  // Visit this node, add the submesh
+  for (unsigned meshIdx = 0; meshIdx < _node->mNumMeshes; ++meshIdx)
+  {
+    auto assimpMeshIdx = _node->mMeshes[meshIdx];
+    auto& assimpMesh = _scene->mMeshes[assimpMeshIdx];
+    auto nodeName = ToString(_node->mName);
+    auto subMesh = this->CreateSubMesh(assimpMesh, _transform);
+    subMesh.SetName(nodeName);
+    // Now add the bones to the skeleton
+    if (assimpMesh->HasBones() && _scene->HasAnimations())
+    {
+      // TODO(luca) merging skeletons here
```

Comment:
Hi @luca-della-vedova,
Can you eloborate on this TODO little bit more, I want to implement it, but I did not quite understand it.